File: lib/hyprdesk/cardspec.py
"""Widget template specs — TOML files in ~/.config/hyprcard/templates/.

A template declares a card: metadata, instantiation params, ONE data
source, and a vertical stack of rows. The row-type list is FROZEN
(plus `repeat`): a widget that doesn't fit gets a new row primitive in
rows.py, never a spec-language feature — if you need an if, you need a
row type.

Dropping a .toml here self-registers the template in the picker and in
Hypr Settings, exactly like dropping a .conf did for the conky fleet.
"""
import re
import tomllib
from pathlib import Path

TEMPLATES_DIR = Path.home() / ".config/hyprcard/templates"

ROW_TYPES = {"title", "text", "keyval", "bar", "sparkline", "graph", "hr",
             "clock", "calgrid", "notesfile", "heatmap"}
NAME_RE = re.compile(r"^[A-Za-z]\w*$")     # must survive widgets.conf keys
PARAM_TYPES = {"string", "int", "choice"}
BUILTIN_SOURCES = {"clock", "stats", "netgraph", "none"}
SUBST_RE = re.compile(r"\{([\w.]+)\}")


class TemplateError(ValueError):
    """Invalid template file — shown on the picker's red tile."""


class Template:
    def __init__(self, name, data, path=None):
        t = data.get("template")
        if not isinstance(t, dict):
            raise TemplateError("missing [template] table")
        self.name = t.get("name", name)
        if not NAME_RE.match(self.name):
            raise TemplateError(
                f"template name {self.name!r} must match [A-Za-z]\\w*")
        if self.name != name:
            raise TemplateError(
                f"template name {self.name!r} != file stem {name!r}")
        self.path = path
        self.label = t.get("label", self.name.title())
        self.icon = t.get("icon", "")
        self.description = t.get("description", "")
        self.multi = bool(t.get("multi_instance", False))
        try:
            self.width = int(t.get("default_width", 300))
        except (TypeError, ValueError) as e:
            raise TemplateError(f"default_width must be an integer: {e}")
        self.default_pos = t.get("default_pos", "top_left")

        self.params = {}
        for pname, p in (data.get("params") or {}).items():
            if not NAME_RE.match(pname):
                raise TemplateError(f"param {pname!r} must match [A-Za-z]\\w*")
            ptype = p.get("type", "string")
            if ptype not in PARAM_TYPES:
                raise TemplateError(f"param {pname}: unknown type {ptype!r}")
            self.params[pname] = {
                "label": p.get("label", pname), "type": ptype,
                "default": str(p.get("default", "")),
                "choices": [str(c) for c in p.get("choices", [])],
            }

        s = data.get("source") or {}
        self.source_cmd = s.get("cmd", "")
        self.source_builtin = s.get("builtin", "")
        try:
            self.interval = max(1, int(s.get("interval", 60)))
        except (TypeError, ValueError) as e:
            raise TemplateError(f"source.interval must be an integer: {e}")
        if self.source_builtin and self.source_builtin not in BUILTIN_SOURCES:
            raise TemplateError(f"unknown builtin source {self.source_builtin!r}")
        if not self.source_cmd and not self.source_builtin:
            self.source_builtin = "none"   # static rows (clock rows self-tick)

        self.rows = data.get("rows") or []
        if not isinstance(self.rows, list) or not self.rows:
            raise TemplateError("template needs at least one [[rows]] entry")
        self._check_rows(self.rows)

        self.preview_fields = parse_fields(
            (data.get("preview") or {}).get("fields", ""))

    def _check_rows(self, rows, depth=0):
        for r in rows:
            rtype = r.get("type")
            if rtype == "repeat":
                if depth:
                    raise TemplateError("repeat cannot nest")
                if not r.get("over"):
                    raise TemplateError("repeat needs over = \"prefix\"")
                self._check_rows(r.get("rows") or [], depth + 1)
            elif rtype not in ROW_TYPES:
                raise TemplateError(f"unknown row type {rtype!r}")


FIELD_MAX = 2000       # a runaway field value must not become a render bomb


def parse_fields(text):
    """key=value lines (key may be dotted: coin.0.name) → flat dict.
    Values are capped defensively at the seam."""
    fields = {}
    for line in str(text).splitlines():
        m = re.match(r"^([\w.]+)\s*=\s*(.*)$", line.strip())
        if m:
            fields[m.group(1)] = m.group(2)[:FIELD_MAX]
    return fields


def subst(text, params, fields=None):
    """Replace {param} / {field.path} in template strings.

    Unresolved refs render as '—' (a fetch that hasn't returned yet must
    degrade gracefully, never show raw braces)."""
    def rep(m):
        key = m.group(1)
        if key in params:
            return params[key]
        if fields and key in fields:
            return fields[key]
        return "—"
    return SUBST_RE.sub(rep, str(text))


def load_template(path):
    """Template from one .toml (raises TemplateError on any problem)."""
    try:
        data = tomllib.loads(Path(path).read_text())
    except (OSError, tomllib.TOMLDecodeError) as e:
        raise TemplateError(str(e)) from e
    return Template(Path(path).stem, data, path=str(path))


def load_templates():
    """{name: Template | TemplateError} for every file in the dir —
    broken templates surface as errors, never as silent absence."""
    out = {}
    if TEMPLATES_DIR.is_dir():
        for p in sorted(TEMPLATES_DIR.glob("*.toml")):
            try:
                out[p.stem] = load_template(p)
            except TemplateError as e:
                out[p.stem] = e
            except Exception as e:      # belt-and-braces: a broken template
                out[p.stem] = TemplateError(str(e))   # must NEVER kill the host
    return out


def instances(c):
    """[(instance_id, template_name)] from widgets.conf inst_* keys."""
    out = []
    for k, v in c.items():
        if k.startswith("inst_") and NAME_RE.match(k[5:]):
            out.append((k[5:], v))
    return sorted(out)


def instance_params(c, inst_id, template):
    """Resolved params for an instance: <id>_p_<name> keys over template
    defaults. (Legacy globals github_user/trading_coins/dev_repos are
    migrated into _p_ keys at startup by the host, then never consulted —
    a live fallback here leaked old values back when the picker CLEARED a
    Configure field.)"""
    out = {}
    for pname, p in template.params.items():
        val = c.get(f"{inst_id}_p_{pname}")
        out[pname] = str(val) if val is not None else p["default"]
    return out


LEGACY_PARAM_KEYS = {"github_user": "user", "trading_coins": "coins",
                     "dev_repos": "repos"}


def migrate_legacy_params(c, templates):
    """One-shot: fold legacy global param keys into the matching
    instances' <id>_p_<name> keys — only where the instance's template
    actually declares that param and has no explicit value yet. Returns
    {changes}, {} if nothing to do."""
    changes = {}
    for inst_id, tname in instances(c):
        t = templates.get(tname)
        if not isinstance(t, Template):
            continue
        for legacy_key, pname in LEGACY_PARAM_KEYS.items():
            if (legacy_key in c and pname in t.params
                    and f"{inst_id}_p_{pname}" not in c):
                changes[f"{inst_id}_p_{pname}"] = c[legacy_key]
    return changes

File: rust/hyprdesk/src/cardspec.rs
//! Widget template specs — TOML files in ~/.config/hyprcard/templates/,
//! ported from lib/hyprdesk/cardspec.py. The row-type list is FROZEN
//! (plus `repeat`): a widget that doesn't fit gets a new row primitive,
//! never a spec-language feature — if you need an if, you need a row type.
//!
//! `examples/cardspec_parity.rs` loads the REAL template dir and the REAL
//! conf through both implementations and diffs every derived value.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

pub const ROW_TYPES: [&str; 11] = [
    "title", "text", "keyval", "bar", "sparkline", "graph", "hr", "clock", "calgrid",
    "notesfile", "heatmap",
];
pub const BUILTIN_SOURCES: [&str; 4] = ["clock", "stats", "netgraph", "none"];
const FIELD_MAX: usize = 2000; // a runaway field must not become a render bomb

fn name_ok(s: &str) -> bool {
    let mut ch = s.chars();
    matches!(ch.next(), Some(c) if c.is_ascii_alphabetic())
        && ch.all(|c| c.is_alphanumeric() || c == '_')
}

fn templates_dir() -> PathBuf {
    crate::home().join(".config/hyprcard/templates")
}

#[derive(Debug, Clone)]
pub struct TemplateError(pub String);

#[derive(Debug, Clone)]
pub struct Param {
    pub label: String,
    pub ptype: String,
    pub default: String,
    pub choices: Vec<String>,
}

/// One row spec, kept as raw TOML — the renderers read ad-hoc keys and a
/// typed struct per row type would freeze what the python leaves open.
pub type RowSpec = toml::Table;

#[derive(Debug, Clone)]
pub struct Template {
    pub name: String,
    pub path: Option<String>,
    pub label: String,
    pub icon: String,
    pub description: String,
    pub multi: bool,
    pub width: i32,
    pub default_pos: String,
    pub action: String,
    pub params: HashMap<String, Param>,
    pub source_cmd: String,
    pub source_builtin: String,
    pub interval: u64,
    pub rows: Vec<RowSpec>,
    pub preview_fields: HashMap<String, String>,
}

fn s(v: Option<&toml::Value>, default: &str) -> String {
    match v {
        Some(toml::Value::String(x)) => x.clone(),
        Some(other) => other.to_string(),
        None => default.to_string(),
    }
}

fn check_rows(rows: &[RowSpec], depth: usize) -> Result<(), TemplateError> {
    for r in rows {
        let rtype = r.get("type").and_then(|v| v.as_str()).unwrap_or("");
        if rtype == "repeat" {
            if depth > 0 {
                return Err(TemplateError("repeat cannot nest".into()));
            }
            if r.get("over").and_then(|v| v.as_str()).unwrap_or("").is_empty() {
                return Err(TemplateError("repeat needs over = \"prefix\"".into()));
            }
            let inner: Vec<RowSpec> = r
                .get("rows")
                .and_then(|v| v.as_array())
                .map(|a| a.iter().filter_map(|v| v.as_table().cloned()).collect())
                .unwrap_or_default();
            check_rows(&inner, depth + 1)?;
        } else if !ROW_TYPES.contains(&rtype) {
            return Err(TemplateError(format!("unknown row type {rtype:?}")));
        }
    }
    Ok(())
}

impl Template {
    pub fn parse(name: &str, data: &toml::Table, path: Option<&Path>) -> Result<Template, TemplateError> {
        let t = data
            .get("template")
            .and_then(|v| v.as_table())
            .ok_or_else(|| TemplateError("missing [template] table".into()))?;
        let tname = s(t.get("name"), name);
        if !name_ok(&tname) {
            return Err(TemplateError(format!(
                "template name {tname:?} must match [A-Za-z]\\w*"
            )));
        }
        if tname != name {
            return Err(TemplateError(format!(
                "template name {tname:?} != file stem {name:?}"
            )));
        }
        let width = match t.get("default_width") {
            None => 300,
            Some(toml::Value::Integer(i)) => *i as i32,
            Some(other) => {
                return Err(TemplateError(format!(
                    "default_width must be an integer: {other}"
                )))
            }
        };

        let mut params = HashMap::new();
        if let Some(ps) = data.get("params").and_then(|v| v.as_table()) {
            for (pname, p) in ps {
                if !name_ok(pname) {
                    return Err(TemplateError(format!(
                        "param {pname:?} must match [A-Za-z]\\w*"
                    )));
                }
                let p = p.as_table().cloned().unwrap_or_default();
                let ptype = s(p.get("type"), "string");
                if !["string", "int", "choice"].contains(&ptype.as_str()) {
                    return Err(TemplateError(format!(
                        "param {pname}: unknown type {ptype:?}"
                    )));
                }
                params.insert(
                    pname.clone(),
                    Param {
                        label: s(p.get("label"), pname),
                        ptype,
                        default: s(p.get("default"), ""),
                        choices: p
                            .get("choices")
                            .and_then(|v| v.as_array())
                            .map(|a| a.iter().map(|c| s(Some(c), "")).collect())
                            .unwrap_or_default(),
                    },
                );
            }
        }

        let src = data
            .get("source")
            .and_then(|v| v.as_table())
            .cloned()
            .unwrap_or_default();
        let source_cmd = s(src.get("cmd"), "");
        let mut source_builtin = s(src.get("builtin"), "");
        let interval = match src.get("interval") {
            None => 60,
            Some(toml::Value::Integer(i)) => (*i).max(1) as u64,
            Some(other) => {
                return Err(TemplateError(format!(
                    "source.interval must be an integer: {other}"
                )))
            }
        };
        if !source_builtin.is_empty() && !BUILTIN_SOURCES.contains(&source_builtin.as_str()) {
            return Err(TemplateError(format!(
                "unknown builtin source {source_builtin:?}"
            )));
        }
        if source_cmd.is_empty() && source_builtin.is_empty() {
            source_builtin = "none".into(); // static rows (clock rows self-tick)
        }

        let rows: Vec<RowSpec> = data
            .get("rows")
            .and_then(|v| v.as_array())
            .map(|a| a.iter().filter_map(|v| v.as_table().cloned()).collect())
            .unwrap_or_default();
        if rows.is_empty() {
            return Err(TemplateError(
                "template needs at least one [[rows]] entry".into(),
            ));
        }
        check_rows(&rows, 0)?;

        let preview_fields = parse_fields(&s(
            data.get("preview")
                .and_then(|v| v.as_table())
                .and_then(|p| p.get("fields")),
            "",
        ));

        Ok(Template {
            name: tname.clone(),
            path: path.map(|p| p.display().to_string()),
            label: s(t.get("label"), &title_case(&tname)),
            icon: s(t.get("icon"), ""),
            description: s(t.get("description"), ""),
            multi: t.get("multi_instance").and_then(|v| v.as_bool()).unwrap_or(false),
            width,
            default_pos: s(t.get("default_pos"), "top_left"),
            action: s(t.get("action"), "").trim().to_string(),
            params,
            source_cmd,
            source_builtin,
            interval,
            rows,
            preview_fields,
        })
    }
}

/// python str.title() for the label default ("netgraph" → "Netgraph").
fn title_case(sv: &str) -> String {
    let mut out = String::new();
    let mut boundary = true;
    for c in sv.chars() {
        if c.is_alphanumeric() {
            if boundary {
                out.extend(c.to_uppercase());
            } else {
                out.extend(c.to_lowercase());
            }
            boundary = false;
        } else {
            out.push(c);
            boundary = true;
        }
    }
    out
}

/// key=value lines (key may be dotted: coin.0.name) → flat map.
pub fn parse_fields(text: &str) -> HashMap<String, String> {
    let mut out = HashMap::new();
    for line in text.lines() {
        let line = line.trim();
        let Some(eq) = line.find('=') else { continue };
        let key = line[..eq].trim();
        if key.is_empty()
            || !key
                .chars()
                .all(|c| c.is_alphanumeric() || c == '_' || c == '.')
        {
            continue;
        }
        let val = line[eq + 1..].trim_start();
        out.insert(
            key.to_string(),
            val.chars().take(FIELD_MAX).collect::<String>(),
        );
    }
    out
}

/// Replace {param} / {field.path}; unresolved refs render as '—'.
pub fn subst(
    text: &str,
    params: &HashMap<String, String>,
    fields: Option<&HashMap<String, String>>,
) -> String {
    let mut out = String::new();
    let mut rest = text;
    while let Some(open) = rest.find('{') {
        out.push_str(&rest[..open]);
        let tail = &rest[open + 1..];
        if let Some(close) = tail.find('}') {
            let key = &tail[..close];
            if !key.is_empty()
                && key
                    .chars()
                    .all(|c| c.is_alphanumeric() || c == '_' || c == '.')
            {
                if let Some(v) = params.get(key) {
                    out.push_str(v);
                } else if let Some(v) = fields.and_then(|f| f.get(key)) {
                    out.push_str(v);
                } else {
                    out.push('—');
                }
                rest = &tail[close + 1..];
                continue;
            }
        }
        // no closing brace / not a ref: keep the brace literally
        out.push('{');
        rest = tail;
    }
    out.push_str(rest);
    out
}

pub fn load_template(path: &Path) -> Result<Template, TemplateError> {
    let text = std::fs::read_to_string(path).map_err(|e| TemplateError(e.to_string()))?;
    let data: toml::Table = toml::from_str(&text).map_err(|e| TemplateError(e.to_string()))?;
    let stem = path
        .file_stem()
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_default();
    Template::parse(&stem, &data, Some(path))
}

/// {name: Ok(Template) | Err(TemplateError)} — broken templates surface as
/// errors, never as silent absence.
pub fn load_templates() -> HashMap<String, Result<Template, TemplateError>> {
    let mut out = HashMap::new();
    let Ok(dir) = std::fs::read_dir(templates_dir()) else {
        return out;
    };
    let mut paths: Vec<PathBuf> = dir
        .flatten()
        .map(|e| e.path())
        .filter(|p| p.extension().is_some_and(|e| e == "toml"))
        .collect();
    paths.sort();
    for p in paths {
        let stem = p
            .file_stem()
            .map(|s| s.to_string_lossy().to_string())
            .unwrap_or_default();
        out.insert(stem, load_template(&p));
    }
    out
}

/// [(instance_id, template_name)] from widgets.conf inst_* keys, sorted.
pub fn instances(c: &HashMap<String, String>) -> Vec<(String, String)> {
    let mut out: Vec<(String, String)> = c
        .iter()
        .filter(|(k, _)| k.starts_with("inst_") && name_ok(&k[5..]))
        .map(|(k, v)| (k[5..].to_string(), v.clone()))
        .collect();
    out.sort();
    out
}

/// Resolved params for an instance: <id>_p_<name> keys over defaults.
pub fn instance_params(
    c: &HashMap<String, String>,
    inst_id: &str,
    template: &Template,
) -> HashMap<String, String> {
    template
        .params
        .iter()
        .map(|(pname, p)| {
            let v = c
                .get(&format!("{inst_id}_p_{pname}"))
                .cloned()
                .unwrap_or_else(|| p.default.clone());
            (pname.clone(), v)
        })
        .collect()
}

const LEGACY_PARAM_KEYS: [(&str, &str); 3] = [
    ("github_user", "user"),
    ("trading_coins", "coins"),
    ("dev_repos", "repos"),
];

/// One-shot fold of legacy global params into <id>_p_<name> keys.
pub fn migrate_legacy_params(
    c: &HashMap<String, String>,
    templates: &HashMap<String, Result<Template, TemplateError>>,
) -> HashMap<String, String> {
    let mut changes = HashMap::new();
    for (inst_id, tname) in instances(c) {
        let Some(Ok(t)) = templates.get(&tname) else {
            continue;
        };
        for (legacy_key, pname) in LEGACY_PARAM_KEYS {
            let target = format!("{inst_id}_p_{pname}");
            if c.contains_key(legacy_key)
                && t.params.contains_key(pname)
                && !c.contains_key(&target)
            {
                changes.insert(target, c[legacy_key].clone());
            }
        }
    }
    changes
}
